capsules/src/pwm.rs

```diff
@@ -0,0 +1,137 @@
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::OptionalCell;
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Pwm as usize;
+
+// An empty app, for potential uses in future updates of the driver
+#[derive(Default)]
+pub struct App;
+
+pub struct Pwm<'a, const NUM_PINS: usize> {
+    /// The usable pwm pins.
+    pwm_pins: &'a [&'a dyn hil::pwm::PwmPin; NUM_PINS],
+    /// Per-app state.
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    /// An array of apps associated to their reserved pins.
+    active_app: [OptionalCell<ProcessId>; NUM_PINS],
+}
+
+impl<'a, const NUM_PINS: usize> Pwm<'a, NUM_PINS> {
+    pub fn new(
+        pwm_pins: &'a [&'a dyn hil::pwm::PwmPin; NUM_PINS],
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> Pwm<'a, NUM_PINS> {
+        assert!(NUM_PINS <= (u16::MAX as usize));
+        const EMPTY: OptionalCell<ProcessId> = OptionalCell::empty();
+        Pwm {
+            pwm_pins: pwm_pins,
+            apps: grant,
+            active_app: [EMPTY; NUM_PINS],
+        }
+    }
+
+    pub fn is_valid_app(&self, processid: ProcessId, pin: usize) -> bool {
```

Comment:
Ah -- I think this method is a bit confusing. I assumed it was checking the validity of the application (e.g., is running), rather than whether the app can acquire the pin. I'd suggest changing this to `claim_pin`, and add a `release_pin`.